### ISSUE TITLE
Allow plot methods to register custom alt text

### ISSUE THREAD

**Issue by hadley** [yihui/knitr#2001]
It would be useful if plot methods could register default alt text to use when `fig.alt` is not supplied. I'm not sure how exactly this would work, but maybe knitr could provide some function that plot methods could call. Then (e.g.) `print.ggplot2` could do:

```R
print.ggplot2 <- function(x, ...) {
  if (isTRUE(getOption("knitr.in.progress")) && is_installed("knitr", 2.0.0)) {
    knitr::set_default_alt(x$labels$title)
  }
  ...
}
```

This would be a default, so would be overridden by explicitly supplied `fig.alt`. (And ggplot2 would also gain some richer ways of setting the alt text, probably via `labs()`).
 

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by cderv** [comment]
I was curious to go through an idea following the discussion we had. Here some notes on how this could work

Currently, I don't think **knitr** will know specifically about **ggplot2**. After evaluation, I believe **knitr** will see the result as a `recordedplot` and won't be able to retrieve any specific object attributes set by ggplot. So I am assuming first that we want to modify an option during the evaluation of a chunk. 

The idea here is to have an external code modifies a knitr's chunk option during evaluation for the rest of the code chunk processing. This could be either a mechanism dedicated to a specific option like `fig.alt` or for any chunk option. 

For the later, one idea could be to use `opts_current$set(fig.alt = )` to change the current option of a chunk during evaluation of a code chunk (or rather in `knit_print()` phase). Something like this: 

````markdown
---
output: 
  html_document:
    keep_md: true
---

# Setup 

```{r}
library(knitr)
knit_print.ggplot <- function(x, options, ...) {
  if (!is.null(x$labels$alt) && is.null(options$fig.alt)) {
    knitr::opts_current$set(fig.alt = x$labels$alt)
  }
  NextMethod()
}

.S3method("knit_print", "ggplot", knit_print.ggplot)
```


# Plot with alt text

```{r}
library(ggplot2)
ggplot(mpg, aes(displ, hwy, colour = class)) + 
  geom_point() + 
  labs(alt = "An alt text")
```

````

However, this currently does not work because options are considered immutable by external evaluated code. This would require to update `opts_current` by any modification knitr is doing on option, and allow modification during code chunk evaluation. This would mean setting `opts_current$set(options)` before `evaluate` is called and call `options = opts_current$get()` after. 

```diff
diff --git a/R/block.R b/R/block.R
index 1d9b5788..0dd86673 100644
--- a/R/block.R
+++ b/R/block.R
@@ -224,10 +224,11 @@ eng_r = function(options) {
   if (keep != 'none' && is.null(options$fig.ext))
     options$fig.ext = dev2ext(options$dev)

   cache.exists = cache$exists(options$hash, options$cache.lazy)
   evaluate = knit_hooks$get('evaluate')
+  # save option as they were modified since last save to opts_current
+  opts_current$set(options)
   # return code with class 'source' if not eval chunks
   res = if (is_blank(code)) list() else if (isFALSE(ev)) {
     as.source(code)
   } else if (cache.exists && isFALSE(options$cache.rebuild)) {
     fix_evaluate(cache$output(options$hash, 'list'), options$cache == 1)
@@ -241,10 +242,11 @@ eng_r = function(options) {
         if (options$error && options$include) 0L else 2L
       },
       output_handler = knit_handlers(options$render, options)
     )
   )
+  # set options to possible modified values during evaluation
+  options = opts_current$get()
   if (options$cache %in% 1:2 && (!cache.exists || isTRUE(options$cache.rebuild))) {
     # make a copy for cache=1,2; when cache=2, we do not really need plots
     res.orig = if (options$cache == 2) remove_plot(res, keep == 'high') else res
   }
```

This allows to have the `fig.alt` modification taken into account, and would allow any other package to modify some options if necessary.

A more targeted approach would be to determine a list of options that could be modified, store them in a new special knitr list object, and always retrieve them from there when used. 

````r
# set by default in knitr
opts_mutable = new_defaults(opts_chunk$get(c("fig.alt")))
# would be used in external code
opts_mutable$set(fig.alt = "Alt text") 
# maybe though a wrapper 
change_chunk_opts <- function(...) {
    check_is_mutable(...)
	opts_mutable$set(...)
}

# knitr would use this when needing the option
opts_mutable$get("fig.alt")
```

But more questions: 
* Should `opts_current` be updated also ? 
* What is the advantage of this more specific approach ? Drawback: it kind of duplicate `opts_current`

I leave that here for now and think of another possible approach if any.

**Comment by hadley** [comment]
I don't think this can be an option (since it would be very hard to reassemble the alt text provided by the user and the alt package provided by a plotting package); I think we'd want some other knitr call that accumulated alt text across plots in a chunk and then combined with user supplied options.

**Comment by cderv** [comment]
That is helpful, Thanks. `fig.alt` can handle a vector (and recycling) and I had that in mind for the multiple plot scenario. I'll think differently about this and see where it leads.

**Comment by hadley** [comment]
I was thinking you might want to do `fig.alt = c(NA, "custom alt text")`.

**Comment by hadley** [comment]
I realised that shiny will have the same need to capture the alt text, so we might also need something a bit more generic.
